specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1258,591 @@ int __cdecl wmain(int argc, WCHAR** argv)
+                          - position: 0.5
+                          - color: { 1, 1, 1, 1 }
+                        D2D1_GRADIENT_STOP:
+                          - position: 1
+                          - color: { 1, 0, 0, 1 }
+```
+
+For convenience, the sample defines `Indent` and `PropName` helper types with assocaited stream
+output opreators. It also defines stream output operators for various API types. These helper types
+and operators are shown at the end of this section.
+
+The `DumpPaintTree` function creates an `IDWritePaintReader` object, sets the current glyph, outputs
+the glyph properties, and then calls the recursive `DumpPaintElement` function to output the tree of
+paint elements.
+
+```c++
+void DumpPaintTree(std::ostream& out, IDWriteFontFace7* fontFace, uint32_t glyphIndex)
+{
+    wil::com_ptr<IDWritePaintReader> paintReader;
+    THROW_IF_FAILED(fontFace->CreatePaintReader(
+        DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE,
+        DWRITE_PAINT_FEATURE_LEVEL_COLR_V1,
+        &paintReader
+    ));
+
+    DWRITE_PAINT_ELEMENT paintElement;
+    D2D_RECT_F clipBox;
+    DWRITE_PAINT_ATTRIBUTES attributes;
+    THROW_IF_FAILED(paintReader->SetCurrentGlyph(glyphIndex, &paintElement, &clipBox, &attributes));
+
+    out << PropName{ "glyphIndex" } << glyphIndex << '\n'
+        << PropName{ "clipBox" } << clipBox << '\n'
+        << PropName{ "attributes" } << attributes << '\n'
+        << PropName{ "paintTree" } << '\n';
+    DumpPaintElement(out, Indent{ 1 }, paintReader.get(), paintElement);
+}
+```
 
-## DWRITE_FACTORY_TYPE Enumeration
+The recursive `DumpPaintElement` function outputs a description of the specified paint element and
+its children. The `DWRITE_PAINT_ELEMENT` structure is a tagged union. The `DumpPaintElement`
+branches on its `paintType` member to determine what to write.
+
+```c++
+void DumpPaintElement(std::ostream& out, Indent indent, IDWritePaintReader* reader, DWRITE_PAINT_ELEMENT& element)
+{
+    // Helper to write gradient information, for gradient paint types.
+    auto WriteGradient = [&](D2D1_EXTEND_MODE extendMode, uint32_t gradientStopCount)
+    {
+        std::vector<D2D1_GRADIENT_STOP> stops;
+        stops.resize(gradientStopCount);
+        THROW_IF_FAILED(reader->GetGradientStops(0, gradientStopCount, /*out*/ stops.data()));
+
+        out << indent << PropName{ "extendMode" } << extendMode << '\n';
+        out << indent << PropName{ "gradientStops" } << '\n';
+        indent.value++;
+        for (auto& stop : stops)
+        {
+            out << indent << "D2D1_GRADIENT_STOP:\n"
+                << indent << PropName{ "position" } << stop.position << '\n'
+                << indent << PropName{ "color" } << stop.color << '\n';
+        }
+        indent.value--;
+    };
+
+    // Helper to recursively call DumpPaintElement for a child paint element.
+    auto Recurse = [&]()
+    {
+        DumpPaintElement(out, Indent{ indent.value + 1 }, reader, element);
```

Comment:
NIT: Recommend considering some sort of "AutoIndent" class which indents/unindents based on constructor/destructor. Will avoid the need for all the +1/-1 indent logic.

Reply:
I looked into this, but I think it would require adding more boilerplate code than it's worth, especially as it's all peripheral to the APIs being demonstrated. I've modified the sample to hopefully be a little clearer. It now passes `indent + 1` at a couple of function call sites but never modifies indent within the function.